Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/AllowedUndeclaredReadsTests.cs

```diff
@@ -383,6 +383,66 @@ public void ReadViolationsAreDetectedEvenWhenRunningFromCache()
             AssertErrorEventLogged(LogEventId.DependencyViolationWriteInUndeclaredSourceRead);
         }
 
+        /// <summary>
+        /// TODO: blocking writes on existing undeclared inputs is not implemented on Mac yet
+        /// </summary>
+        [FactIfSupported(requiresWindowsBasedOperatingSystem: true)]

```

Comment:
Can you summarize why this functionality needs to be implemented inside the sandbox instead of as part of the engine?

Couldn't you instead enumerate the content of all shared opaque directories (after they've been scrubbed of shared opaque outputs), remember those files as source files, and then inside the engine issue a DFA for every undeclared write to any of those source files?  You could even add those source files explicitly to the FileAccessManifest and disallow writes there.

Reply:
Enumerating arbitrary cones before each pip runs is not an option perf-wise. Shared opaques can be arbitrarily big. E.g. for the MSBuild case, shared opaques are usually the entire repo, including a mix of sources and outputs.

---
In reply to: [299174940](https://github.com/microsoft/BuildXL/pull/536#discussion_r299174940) [](ancestors = 299174940)

Reply:
I meant: enumerate them once after scrubbing (for scrubbing you have to enumerate them anyway). 

Enumerating once before the build instead of before every pip should be enough, because, I believe, we already detect the case when 2 different pips end up writing the same undeclared file (via their respective shared opaque directories)

Reply:
Medium-term we want to get rid of scrubbing for shared opaques (or at least minimize it to some unavoidable set of cases). E.g. when Helium is present, scrubbing can be removed altogether. So I wouldn't like to pile up on that feature. 
I'm also concerned about generating a 'preventive' detours manifest for every pip that is probably going to be many orders of magnitude bigger than what we have today. But still my main reason for not going that route is the first argument.

---
In reply to: [299208300](https://github.com/microsoft/BuildXL/pull/536#discussion_r299208300) [](ancestors = 299208300)